camayoc/qcs_models.py

```diff
@@ -42,43 +65,140 @@ def __ne__(self, other):
+
+        :param ``**kwargs``: Additional arguments accepted by Requests's
+            `request.request()` method.
+
+        Returns a requests.models.Response. A successful delete has the return
+        code `204`.
+        """
+        return self.client.delete(self.path(), **kwargs)
+
+    def assert_matches_server(self):
+        """Assert that the data on the server matches this object."""
+        other = self.read_self().json()
+        assert self == other
+
 
 class HostCredential(QCSObject):
-    """A class to aid in CRUD test cases for host credentials."""
+    """A class to aid in CRUD tests of Host Credentials on the QCS server.
+
+    Host credentials can be created by instantiating a HostCredential
+    object. A unique name and username are provided by default.
+    In order to create a valid host credential you must specify either a
+    password or ssh_keyfile.
+
+    Example::
+        >>> from camayoc import api
+        >>> from camayoc.qcs_models import HostCredential
+        >>> client = api.QCSClient()
+        >>> cred = HostCredential(password='foo')
+        >>> # The create method automatically sets the credential's `_id`
+        >>> cred.create()
+        >>> actual_cred = cred.read_self().json()
+        >>> # the __eq__() method is customized to
+        >>> # allow == comparison between the json returned by the server
+        >>> # and the HostCredential object used to create it.
+        >>> assert actual_cred == cred
+    """
 
     def __init__(
             self,
+            client=None,
             name=None,
             username=None,
             password=None,
             ssh_keyfile=None,
             sudo_password=None,
             _id=None):
-        """Host Credential objects encapsulate host credential data.
-
-        Host credentials can be created by instantiating a HostCredential
-        object. A unique name and username are provided by default.
-        In order to create a valid host credential you must specify either a
-        password or ssh_keyfile.
-
-        Example::
-            >>> from camayoc import api
-            >>> from camayoc.qcs_models import HostCredential
-            >>> client = api.QCSClient()
-            >>> host = HostCredential(password='foo')
-            >>> create_response = client.create_credential(host.payload())
-            >>> host._id = create_response.json()['id']
+        """Create a host credential with given data.
+
+        If no arguments are passed, then a api.Client will be initialized and a
+        uuid4 generated for the name and username.
+
+        For a HostCredential to be successfully created on the QCS server,
+        a password XOR a ssh_keyfile must be provided.
         """
-        super().__init__(name=name, _id=_id)
+        super().__init__(client=client, name=name, _id=_id)
+        self.endpoint = QCS_CREDENTIALS_PATH
         self.username = str(uuid.uuid4()) if not username else username
```

Comment:
Taking a second look on this line (even though it is not related to this PR) this will not allow us to submit empty username to the server so we can't check if it will raise a validation error.

Maybe checking if name is not None is better here. So it allows empty username (`''` string) but will generate one for user when it is None.

Reply:
I see what you are saying, because '' is also false, so even if I stated `name=''` explicitly when I created it this would end up with a uuid4 name.

Reply:
Yeah, you got it